File: samples-dev/plugins/src/main/java/org/gradle/samples/plugins/generators/SamplesExtension.java
package org.gradle.samples.plugins.generators;

import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.Project;

public class SamplesExtension {
    private final NamedDomainObjectContainer<ExternalRepo> externalRepos;
    private final NamedDomainObjectContainer<Sample> samples;

    public SamplesExtension(Project project) {
        externalRepos = project.container(ExternalRepo.class, name -> {
            return project.getObjects().newInstance(ExternalRepo.class, name);
        });
        samples = project.container(Sample.class, name -> {
            return project.getObjects().newInstance(Sample.class, name);
        });
    }

    public NamedDomainObjectContainer<ExternalRepo> getExternalRepos() {
        return externalRepos;
    }

    public NamedDomainObjectContainer<Sample> getSamples() {
        return samples;
    }
}

File: samples-dev/plugins/src/main/kotlin/org/gradle/samples/plugins/generators/SamplesExtension.kt
package org.gradle.samples.plugins.generators

import org.gradle.api.Action
import org.gradle.api.Project
import org.gradle.api.model.ObjectFactory
import org.gradle.samples.plugins.SampleGeneratorTask
import javax.inject.Inject


open class SamplesExtension(project: Project) {
    val externalRepos = project.container(ExternalRepo::class.java) { name ->
        project.objects.newInstance(ExternalRepo::class.java, name, project.objects)
    }
}

open class ExternalRepo @Inject constructor(val name: String, objectFactory: ObjectFactory) {
    val repoUrl = objectFactory.property(String::class.java)
    val sourceActions = ArrayList<Action<in SampleGeneratorTask>>()
    val repoActions = ArrayList<Action<in Changes>>()

    /**
     * Adds an action to run to configure the source templates for this repo. The source templates are applied to the Git repo clone
     */
    fun copySource(action: Action<in SampleGeneratorTask>) {
        sourceActions.add(action)
    }

    /**
     * Adds an action to run to apply a change to the repo. These are applied after the source templates, in order added.
     */
    fun change(action: Action<in Changes>) {
        repoActions.add(action)
    }
}